Support reading different barcode types in develop branch

jojastahl opened the pull request:
I added support to read different barcode types. Default is QR Code.
Types can be configured by codeTypes property like this:

```html
<zxing-scanner [codeTypes]="'any'"></zxing-scanner>
```

or

```html
<zxing-scanner [codeTypes]="['oned', 'qr']"></zxing-scanner>
```

odahcam commented:
First thanks! That's a great contribution. 😄 

Could we instead of using this:

```typescript
export type CodeTypes = 'any' | 'oned' | 'qr' | 'code128' | 'itf' | 'ean' | 'datamatrix';
```

Just pass the type directly from `@zxing/library`? That would make the entire work about handling the scanner type much easier, specially on `setCodeTypes`.

@werthdavid I  would like to see your thoughts on that. 🙂

jojastahl commented:
I think it would be possible, but the `BarcodeFormat` does not contain something like `any` or all `oned`. How would we do that?

Is it possible to access these constants from HTML markup? Would we need to write something like

`<zxing-scanner [codeTypes]="[BarcodeFormat.CODE_128, BarcodeFormat.QR_CODE]">`

How html knows about `BarcodeFormat`? Would it be necessary to import this in the component.ts file?

odahcam commented:
> I think it would be possible, but the BarcodeFormat does not contain something like any or all oned. How would we do that?

The MultiFormat reader: https://github.com/zxing-js/library/blob/develop/src/core/MultiFormatReader.ts

> Is it possible to access these constants from HTML markup? Would we need to write something like

Yas, more or less like that I believe. As long as we'r using TypeScript, typing things should be a priority, I'm not confortable with strings there, if we can't make something else better, we should at least match the barcode names with the available formats.

> How html knows about BarcodeFormat? Would it be necessary to import this in the component.ts file?

You can import the readers on the component and pass'em trough variables to the template or configure it directly in the component.

----

As this is a 2.0 targeted feature, I'd like to discuss well what we gonna do please don't feel like I didn't appreciate your work. 🙂

jojastahl commented:
> You can import the readers on the component and pass'em trough variables to the template or configure it directly in the component.

I don't think passing the reader classes (`Code128Reader`, `QRCodeReader`, ...) is a good idea. The library offers a `MultiFormatReader` which is targeted to this usage, as you indicated, and I prefer to use it in all cases. `MultiFormatReader` is customized with hints: `DecodeHintType.POSSIBLE_FORMATS`. These hints use `BarcodeFormat` constants. So your idea to use these constants directly is a good one.

I'm trying now to rework the PR to use the `BarcodeFormat` constants directly instead of strings and facing serveral problems, which I would like to discuss with you:

**1. `BarcodeFormat` doesn't currently have a concept of format groups:** all, all 1D, all 2D. To work around this, I added 
```typescript
const enum BarcodeFormatGroup {
    ALL = -1,
    ONE_D = -2,
    TWO_D = -3
}

type CodeTypes = BarcodeFormat | BarcodeFormatGroup;
```


**2. While point 1. could be a solution, I still have to decode the `BarcodeFormatGroup`** to individual `BarcodeFormat` constants. That forces me to reference all these constants in the `ngx-scanner` to build up something like
```typescript
const OneD: BarcodeFormat[] = [BarcodeFormat.UPC_A, BarcodeFormat.UPC_E, ...];
const TwoD: BarcodeFormat[] = [BarcodeFormat.AZTEC, BarcodeFormat.DATA_MATRIX, ...];
```
As soon as someone would add a new format to the core library, it would be necessary to update these arrays - a solution which I don't like. The only good way I can think of is adding such arrays to the core library, and / or adding even the `BarcodeFormatGroup` to the core library. What do you think on this?


**3. I can't find a way to reference the `BarcodeFormat` or `BarcodeFormatGroup` constants** directly in the html view.
```html
<zxing-scanner [codeTypes]="[BarcodeFormat.UPC_A, BarcodeFormat.DATA_MATRIX]">
</zxing-scanner>
```
The former doesn't work. In order to use enum constants in views the general proposal is to reference the enum in the controller:
```typescript
class AppComponent implements OnInit {
  readonly fmt = BarcodeFormat;
  readonly grp = BarcodeFormatGroup;
  ...
}
```
```html
<zxing-scanner [codeTypes]="[grp.ONE_D, Fmt.DATA_MATRIX]">
</zxing-scanner>
```
Personally, I do not like this. Additionally, this currently doesn't work as BarcodeFormat is a `const enum` which cannot be used like this. We would have to remove the `const` in the library.

IMO, a better way would be to add a property in the controller defining the desired formats and just pass this property:
```typescript
class AppComponent implements OnInit {
  readonly fmts = [BarcodeFormatGroup.ONE_D, BarcodeFormat.DATA_MATRIX];
  ...
}
```
```html
<zxing-scanner [codeTypes]="fmts"></zxing-scanner>
```

Personally, for the sake of simplicity, I would even go further and allow strings in the html again:
```html
<zxing-scanner [codeTypes]="['ONE_D', 'DATA_MATRIX']"></zxing-scanner>
```

For this to work, I propose to turn codeTypes into a property with getter and setter. For the Typescript part, the property would be typed to the enum(s). But in the setter I would handle the case that contrary to the TS rules a string comes in from html and parse it to the enum constants. In order to do this without coding each enum value again, the enums would have to be non-const which would imply to remove the `const` from the `const enum BarcodeFormat` in the library. Then I could just do `BarcodeFormat[the_string]` to resolve the string to its value.


**4. Renaming `codeTypes` to `formats` or `barcodeFormats`** would seem a good idea to me in order to use consistent wording with `BarcodeFormat`.

Uff, a long post. Too long perhaps?
What do you think about all these points?

odahcam commented:
I personally liked the

```json
<zxing-scanner [codeTypes]="['ONE_D', 'DATA_MATRIX']"></zxing-scanner>
``` 

idea, because it joins the usefull with the organizated way.

Also, I think that `formats` is a way better name for that. 😀

I don't think that "all 1D" or "all 2D" matters that much for now. For me we should o without that, release and then, with time, think  in a cool improvement.

odahcam commented:
@jojastahl 

> 1. BarcodeFormat doesn't currently have a concept of format groups: all, all 1D, all 2D. To work around this, I added

About that, you could use/implement the reader from ZXing: `package com.google.zxing.oned.MultiFormatOneDReader;`

jojastahl commented:
I rebased this PR on current `develop` branch and incorporated the discussed modifications.
From the `zxing-js/library` project, this PR needs zxing-js/library#61 to compile, and zxing-js/library#58, zxing-js/library#60 to function properly.

By default, the reader only reads QR codes, as before. But you may now specify which formats to read with attribute `formats`:
```html
<!-- plain attribute -->
<zxing-scanner formats="CODE_39, CODE_128, QR_CODE"></zxing-scanner>
<!-- binding to string -->
<zxing-scanner [formats]="'CODE_39, CODE_128, QR_CODE'"></zxing-scanner>
<!-- binding to string array -->
<zxing-scanner [formats]="['CODE_39', 'CODE_128', 'QR_CODE']"></zxing-scanner>
<!-- binding to controller property -->
<zxing-scanner [formats]="allowedFormats"></zxing-scanner>
```
```typescript
class ... {
  allowedFormats: BarcodeFormat[] = [
    BarcodeFormat.CODE_39, 
    BarcodeFormat.CODE_128, 
    BarcodeFormat.QR_CODE
  ];
  ...
}
```

werthdavid commented:
Looks good! 👍

odahcam commented:
So, shall we focus on these two?

```html
<!-- binding to string array -->
<zxing-scanner [formats]="['CODE_39', 'CODE_128', 'QR_CODE']"></zxing-scanner>
<!-- binding to controller property -->
<zxing-scanner [formats]="allowedFormats"></zxing-scanner>
```

odahcam commented on src/app/modules/zxing-scanner/browser-multi-format-reader.ts:
```diff
@@ -0,0 +1,22 @@
+
+
+export class BrowserMultiFormatReader extends BrowserCodeReader {
+    private multiReader: MultiFormatReader;
```
Let's rename this guy to reader, so it can override it's parent prop, I see no reason to keep both as they have the same value.

jojastahl replied:
I tried this but it's not possible:

```
Class 'BrowserMultiFormatReader' incorrectly extends base class 'BrowserCodeReader'.
  Types have separate declarations of a private property 'reader'.
```

Would be necessary to make `super.reader` `protected` instead of `private` and remove `multiReader` completely. Second problem is it's type. There is one usage where a cast to `MultiFormatReader` would be necessary.

odahcam replied:
I don't see any problem in changing the access level from private to protected. This class is not on ZXing project, so we can design it the way we want. Also, I can't see where this cast would be written. 🤔

jojastahl commented:
> So, `BrowserQRCodeReader` wouldn't be used anymore?

Currently, this is still used as the default, when no `format` is set. But it could easily be removed completely and be replaced by `BrowserMultiFormatReader`.

odahcam commented:
@jojastahl wouldn't this decrease performance?

jojastahl commented:
The overhead should be minimal. When setting the hints, the `MultiFormatReader` builds an array of the requested Reader instances and when decoding, it calls the decode for each reader in the array. In case a single format is specified, this results in a single decode call. The real overhead is walking the array, which is minimal, IMHO. See https://github.com/zxing-js/library/blob/55f400c1b2d03cebee4de83c4c8c960f6a4dd9ed/src/core/MultiFormatReader.ts#L177-L203

However, it is necessary to call decodeWithState, which is what I am doing, and why I override the `decodeBitmap` method.

odahcam commented on src/app/modules/zxing-scanner/browser-multi-format-reader.ts:
```diff
@@ -3,20 +3,17 @@ import { MultiFormatReader, DecodeHintType, BinaryBitmap, Result } from '@zxing/
      */
     protected decodeBitmap(binaryBitmap: BinaryBitmap): Result {
-        return this.multiReader.decodeWithState(binaryBitmap);
+        return (<MultiFormatReader>this.reader).decodeWithState(binaryBitmap);
```
I see, I gonna take a look on that today.

odahcam commented:
@jojastahl please take a look at my last commit, I think we'r ready to go from this point.

jojastahl commented:
👍 This seems to be a good solution.

odahcam commented:
I removed throttling until we find a better solution for changing it, I do not like the one where the scanner needs to be restarted.

Now we should be good to go.
